python/salobj/base_csc.py

```diff
@@ -0,0 +1,336 @@
+# This file is part of salobj.
+#
+# Developed for the LSST Telescope and Site Systems.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["BaseCsc", "State"]
+
+import asyncio
+import enum
+
+from . import utils
+from .controller import Controller
+
+
+class State(enum.IntEnum):
+    """State constants.
+
+    The numeric values come from
+    https://confluence.lsstcorp.org/display/SYSENG/SAL+constraints+and+recommendations
+    """
+    OFFLINE = 4
+    STANDBY = 5
+    DISABLED = 1
+    ENABLED = 2
+    FAULT = 3
+
+
+class BaseCsc:
+    """Base class for a Commandable SAL Component (CSC)
+
+    To implement a CSC in Python define a subclass of this class.
+
+    Parameters
+    ----------
+    sallib : ``module``
+        salpy component library generatedby SAL
+    component_name : `str`
+        Component name and optional index, separated by a colon, e.g.
+        "scheduler" or "Test:2".
+
+    Notes
+    -----
+    The constructor does the following:
+
+    * Create a controller for the component
+    * For each command defined for the component, find a ``do_<name>`` method
+      (which must be present) and add it as a callback to the controller's
+      ``cmd_<name>`` attribute.
+    * The base class provides ``do_<name>`` methods for the standard CSC
+      commands. The default implementation:
+
+        * Checks for validity of the requested state change;
+            if the change is valid then:
+        * Calls ``before_<name>``. This is a no-op in the base class,
+            and is available for the subclass to override.
+        * Changes the state and reports the new value.
+        * Calls ``after_<name>``. Again, this is a no-op in the base class,
+            and is available for the subclass to override.
+        * Report the command as complete
+
+    Standard CSC commands and their associated summary state changes:
+
+    * ``start``: `State.STANDBY` to `State.DISABLED`
+    * ``enable``: `State.DISABLED` to `State.ENABLED`
+
```

Comment:
I said that in the comments but I'll add it here again, can we take a minute to talk about the `OFFLINE` to `STANDBY` transition? Also, it would be good to add some default boiler plate to publish the `settingVersions` event. We can even talk about defining some global library to deal with the settings.